src/v/pandaproxy/schema_registry/sharded_store.cc

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/schema_registry/sharded_store.h"
+
+#include "hashing/jump_consistent_hash.h"
+#include "hashing/xx.h"
+#include "pandaproxy/logger.h"
+#include "pandaproxy/schema_registry/avro.h"
+#include "pandaproxy/schema_registry/errors.h"
+#include "pandaproxy/schema_registry/store.h"
+#include "pandaproxy/schema_registry/types.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/core/std-coroutine.hh>
+
+namespace pandaproxy::schema_registry {
+
+namespace {
+
+ss::shard_id shard_for(const subject& sub) {
+    auto hash = xxhash_64(sub().data(), sub().length());
+    return jump_consistent_hash(hash, ss::smp::count);
+}
+
+ss::shard_id shard_for(schema_id id) {
+    return jump_consistent_hash(id(), ss::smp::count);
+}
+
+} // namespace
+
+ss::future<> sharded_store::start(ss::smp_service_group sg) {
+    _smp_opts = ss::smp_submit_to_options{sg};
+    return _store.start();
+}
+
+ss::future<> sharded_store::stop() { return _store.stop(); }
+
+ss::future<sharded_store::insert_result>
+sharded_store::insert(subject sub, schema_definition def, schema_type type) {
+    auto id = (co_await insert_schema(std::move(def), type)).id;
+    auto [version, inserted] = co_await _store.invoke_on(
+      shard_for(sub), &store::insert_subject, sub, id);
+    co_return insert_result{version, id, inserted};
+}
+
+ss::future<bool> sharded_store::upsert(
+  subject sub,
+  schema_definition def,
+  schema_type type,
+  schema_id id,
+  schema_version version,
+  is_deleted deleted) {
+    co_await upsert_schema(id, std::move(def), type);
+    co_return co_await upsert_subject(sub, version, id, deleted);
+}
+
+ss::future<schema> sharded_store::get_schema(const schema_id& id) {
+    auto schema = co_await _store.invoke_on(
+      shard_for(id), &store::get_schema, id);
+    co_return std::move(schema).value();
+}
+
+ss::future<subject_schema> sharded_store::get_subject_schema(
+  const subject& sub, schema_version version, include_deleted inc_del) {
+    auto v_id = (co_await _store.invoke_on(
+                   shard_for(sub),
+                   &store::get_subject_version_id,
+                   sub,
+                   version,
+                   inc_del))
+                  .value();
+    auto s = co_await get_schema(v_id.id);
```

Comment:
does it make sense to store the data for `get_schema(v_id.id);` on the same core as `shard_for(sub)` to avoid two cross-core messages that are serialized? i'm not exactly sure how the data models map to cores and the dependencies between them. is the distribution purely balancing memory usage?

Reply:
> does it make sense to store the data for `get_schema(v_id.id);` on the same core as `shard_for(sub)` to avoid two cross-core messages that are serialized? i'm not exactly sure how the data models map to cores and the dependencies between them. is the distribution purely balancing memory usage?

A schema can be shared between subjects. So the balancing is primarily for memory, yes.

Reply:
what kind of data volumes are we talking about? we fully replicate all cluster metadata on every core to avoid cross core messaging. it won't work forever, but we also probably won't hit memory limits for quite a while.

Reply:
nm I see you mentioned that they are pretty large down in a later comment.